### PR TITLE
Implement perfect negotiation.

### PR DESCRIPTION
**Pull request by jianjunz**
Implement [perfect negotiation](https://w3c.github.io/webrtc-pc/#perfect-negotiation-example) to avoid collision.

### CODE REVIEW

File: docs/design/perfect_negotiation.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+This document describes how perfect negotiation is implemented in OWT P2P SDK to avoid collision.
+
+**Perfect negotiation**, **polite peer**, **impolite peer** are defined in [Perfect negotiation example of WebRTC 1.0](https://w3c.github.io/webrtc-pc/#perfect-negotiation-example).
+
+## Determining polite peer and impolite peer
+
+OWT client SDKs determines polite peer and impolite peer by comparing client IDs. Sorting these two client IDs alphabetically in increasing order, the first one is polite order, the second one is impolite peer.
+
+Signaling server is required to return a client ID assigned for the client connected after authentication. OWT doesn't define how authentication works and how client IDs are assigned. They depend on application's design. But every client gets a unique client ID after connecting to the signaling server.
```
Comment (taste1981) [inline docs/design/perfect_negotiation.md:11]: assigned to the client.

File: docs/design/perfect_negotiation.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+## Connections
+
+We expect only one `PeerConnection` between two endpoints. A random ID is generated when creating a new `PeerConnection`. All messages (offers, answers, ICE candidates) for this `PeerConnection` have a `connectionId` property with the connection ID as its value.
```
Comment (taste1981) [inline docs/design/perfect_negotiation.md:15]: Please clarify if Client A and Client B sets up the peer connection, will connectionId value to be the same for both?

File: docs/design/perfect_negotiation.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+## Behaviors
+
+When WebRTC collision occurred, it basically follows the perfect negotiation example in WebRTC 1.0. This section only describes the implementation for signaling collision.
```
Comment (taste1981) [inline docs/design/perfect_negotiation.md:19]: occurs

File: docs/design/perfect_negotiation.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+An example is both sides call `send` to create a data channel and send a message to the remote endpoint at the same time. Each side creates a new `PeerConnection`, and a connection ID. These two connection IDs are different. Then each of them will receive signaling messages with connection ID differ from the local one.
+
+### Polite peer
+
+The polite peer is the controlled side. When a signaling message with a new connection ID is received, it stops current PeerConnection， create a new one, and associate it with the remote connection ID.
+
+### Impolite peer
+
+The polite peer is the controlling side. It ignores remote messages conflict with its own state, and continue its process.
```
Comment (taste1981) [inline docs/design/perfect_negotiation.md:31]: conflicting with its own state, and continues its own process.

File: src/sdk/p2p/peerconnection-channel.js [hunk 5]
```diff
@@ -416,6 +416,16 @@ class P2PPeerConnectionChannel extends EventDispatcher {
   _onOffer(sdp) {
     Logger.debug('About to set remote description. Signaling state: ' +
       this._pc.signalingState);
+    if (this._pc.signalingState !== 'stable') {
+      if (this._isPolitePeer) {
+        // Rollback.
+        Logger.debug('Rollback.');
+        this._pc.setLocalDescription();
+      } else {
+        // Ignore this offer.
```
Comment (taste1981) [inline src/sdk/p2p/peerconnection-channel.js:425]: Add log here for debugging purpose.

File: docs/design/perfect_negotiation.md (hunk 1 above)
Comment (jianjunz) [reply docs/design/perfect_negotiation.md]: Fixed.

File: docs/design/perfect_negotiation.md (hunk 2 above)
Comment (jianjunz) [reply docs/design/perfect_negotiation.md]: Fixed. Connection ID is shared by both clients.

File: docs/design/perfect_negotiation.md (hunk 3 above)
Comment (jianjunz) [reply docs/design/perfect_negotiation.md]: Fixed.

File: docs/design/perfect_negotiation.md (hunk 4 above)
Comment (jianjunz) [reply docs/design/perfect_negotiation.md]: Fixed.

File: src/sdk/p2p/peerconnection-channel.js (hunk 5 above)
Comment (jianjunz) [reply src/sdk/p2p/peerconnection-channel.js]: Fixed. Removed comments and added Logger.debug().

**Review by taste1981 — APPROVED** [review]

**Pull request merged**